### ISSUE TITLE
Allowing multi-processor usage on headless (command-line) mode

### ISSUE THREAD

**Issue by pcm32** [CellProfiler/CellProfiler#1183]
Hi there,

I'm trying to modify the code to allow multiprocessor usage when invoking from the command line in headless mode. I manage to redirect the modules run to the ModuleRunner, but I get an error which is probably associated to the JVM not being started and attached (which it is according to my debugging). I get the following error:

```
Intercepted exception while running module
Traceback (most recent call last):
  File "/home/pcm32/Development/CellProfiler/cellprofiler/pipeline.py", line 383, in run
    self.module.run(self.workspace)
  File "/home/pcm32/Development/CellProfiler/cellprofiler/modules/identifyprimaryobjects.py", line 896, in run
    image = workspace.image_set.get_image(image_name)
  File "/home/pcm32/Development/CellProfiler/cellprofiler/measurements.py", line 1602, in get_image
    image = matching_providers[0].provide_image(self)
  File "/home/pcm32/Development/CellProfiler/cellprofiler/modules/loadimages.py", line 3199, in provide_image
    self.get_name(), url=self.get_url())
  File "/home/pcm32/Development/CellProfiler/bioformats/formatreader.py", line 877, in get_image_reader
    rdr = ImageReader(path=path, url=url)
  File "/home/pcm32/Development/CellProfiler/bioformats/formatreader.py", line 590, in __init__
    self.path)
  File "/home/pcm32/Development/CellProfiler/cellprofiler/utilities/jutil.py", line 1494, in make_instance
    klass = get_env().find_class(class_name)
AttributeError: 'NoneType' object has no attribute 'find_class'
Intercepted exception while running module
Traceback (most recent call last):
  File "/home/pcm32/Development/CellProfiler/cellprofiler/pipeline.py", line 383, in run
    self.module.run(self.workspace)
  File "/home/pcm32/Development/CellProfiler/cellprofiler/modules/identifyprimaryobjects.py", line 896, in run
    image = workspace.image_set.get_image(image_name)
  File "/home/pcm32/Development/CellProfiler/cellprofiler/measurements.py", line 1602, in get_image
    image = matching_providers[0].provide_image(self)
  File "/home/pcm32/Development/CellProfiler/cellprofiler/modules/loadimages.py", line 3199, in provide_image
    self.get_name(), url=self.get_url())
  File "/home/pcm32/Development/CellProfiler/bioformats/formatreader.py", line 877, in get_image_reader
    rdr = ImageReader(path=path, url=url)
  File "/home/pcm32/Development/CellProfiler/bioformats/formatreader.py", line 590, in __init__
    self.path)
  File "/home/pcm32/Development/CellProfiler/cellprofiler/utilities/jutil.py", line 1494, in make_instance
    klass = get_env().find_class(class_name)
AttributeError: 'NoneType' object has no attribute 'find_class'
```

The javabridge is being initialised by the bioformats ``start_cellprofiler_jvm()`` before this happens, but still the get_env() returns a NoneType (which I understand it happens when the javabridge has not started and attached).

I would appreciate any hints that would help me sort this out. There is probably something that I should be executing before the:

```python
worker = ModuleRunner(module, workspace, frame, cpprefs.get_headless())
```

Thanks!

**Comment by 0x00b1** [comment]
It’s a shame this wasn’t merged in 2014. Unfortunately, master and fork are too far apart to push the changes upstream. 😠

**Comment by bgriffen** [comment]
When I run the pipeline in CP, it runs in parallel correctly (55 workers). However, when I run it via a subprocess in Python with the following command:

`cellprofiler -r -c -p /path_to_project_file/headless.cpproj -i /path_to_input_images/ -o /path_to_output/ -d cpdone.txt`

it only does each image sequentially, though it sometimes bursts to 55 threads (likely during an internal analysis step (e.g. IdentifyPrimaryObjects). What is the flag to set the number of workers in headless mode? I've looked elsewhere to no avail.  Setting e.g. `MKL_NUM_THREADS=28 OMP_NUM_THREADS=28 OPENBLAS_NUM_THREADS=28` before the `cellprofiler` cmd also does not make it parallel. Ubuntu 16.04, CP 3.1.8. Help is appreciated.

**Comment by bgriffen** [comment]
Any chance, anyone might be able to help? Or should I submit a new issue?

**Comment by bethac07** [comment]
Running headless will only run a single worker at a time- you need to dispatch different jobs via the `-f` `-l` keywords or `-g` keywords.

**Comment by bgriffen** [comment]
OK, understood, thanks. I thought there might be a way to handle it internally. Lastly, typically, do people doing it this way write out each batch to unique folders e.g. batch_1/ etc. then reassemble the *.csv outputs into single files for aggregate analysis (i.e. to become what is normally output during `Export to Spreadsheet` when done interactively)?

**Comment by rikrdo89** [comment]
This is a long-shot, but has there been any implementation for multi-thread/processor options in the newest version of Cell profiler in the command line?
